marker_api/src/sem/ty.rs

```diff
@@ -154,3 +203,183 @@ macro_rules! impl_ty_data {
     };
 }
 use impl_ty_data;
+
+/// This struct describes a trait and its generic arguments. It is used to check
+/// if a semantic type implements a trait.
+///
+/// See [`TyKind::implements_trait`].
+#[derive(Debug)]
+pub struct TestTraitRef {
+    trait_ref: TyDefIdSource,
+    // TODO generics
+    mode: TraitTestMode,
+}
+
+impl TestTraitRef {
+    pub fn builder() -> TestTraitRefBuilder {
+        TestTraitRefBuilder::new()
+    }
+}
+
+/// A builder used to construct a [`TestTraitRef`] instance.
+#[derive(Debug)]
+pub struct TestTraitRefBuilder {
+    trait_ref: Option<TyDefIdSource>,
+    mode: TraitTestMode,
+}
+
+impl TestTraitRefBuilder {
+    fn new() -> Self {
+        Self {
+            trait_ref: None,
+            mode: TraitTestMode::default(),
+        }
+    }
+
+    pub fn trait_from_path(&mut self, path: impl Into<String>) -> &mut Self {
+        self.trait_ref = Some(TyDefIdSource::Path(path.into()));
+        self
+    }
+
+    pub fn trait_from_id(&mut self, id: impl Into<TyDefId>) -> &mut Self {
+        self.trait_ref = Some(TyDefIdSource::Id(id.into()));
+        self
+    }
+
+    pub fn mode(&mut self, mode: TraitTestMode) -> &mut Self {
+        self.mode = mode;
+        self
+    }
+
+    pub fn build(&mut self) -> TestTraitRef {
+        TestTraitRef {
+            trait_ref: self.trait_ref.take().expect("the trait was never set"),
```

Comment:
I think [`buildstructor`](https://docs.rs/buildstructor/latest/buildstructor/) would be handy in this case since it validates such errors at compile time. Anyway, at some point we'll have to have a proc macro dependency in marker_api.

With buildstructor you could even achieve the API like this:

```rust
ctx.test_trait()
   .arg_foo(...)
   .arg_bar(...)
   .test()
```

In this case you won't even need any owned value as an `arg` like a `String`. Although with this approach it's harder to save the args into a variable and basically impossible to put them into some struct or enum because the builder type buildstructor generates is something very complicated and potentially breakable, so treat it as an anonymous type. However, I think this would be good enough